core/Tracker/Response.php

```diff
@@ -161,6 +172,40 @@ private function outputTransparentGif()
         echo base64_decode($transGifBase64);
     }
 
+    /**
+     * Output a custom tracking image
+     *
+     * @param string $customImage The custom image setting specified in the config
+     * @return bool True if the custom image was successfully output, else false
+     */
+    private function outputCustomImage(string $customImage): bool
+    {
+
+        if (strlen($customImage) > 2 && substr($customImage, -2) == '==') {
+
+            $transGifBase64 = $customImage;
+            Common::sendHeader('Content-Type: image/gif');
+
+            echo base64_decode($transGifBase64);
+            return true;
+        }
+
+        if (file_exists($customImage)) {
+
+            $fp = fopen($customImage, 'rb');
+            $size = getimagesize($customImage); // imagesize is used to get the mime type
```

Comment:
see https://www.php.net/manual/en/function.getimagesize.php

> This function expects filename to be a valid image file. If a non-image file is supplied, it may be incorrectly detected as an image and the function will return successfully, but the array may contain nonsensical values.
> Do not use getimagesize() to check that a given file is a valid image. Use a purpose-built solution such as the Fileinfo extension instead.

My security concern is that someone could configure to print a file that is made to look like an image when it's maybe not. A sys admin would likely not configure anything malicious there but maybe if a different security bug allowed to write certain config files then an attacker might try this way to get various different files printed to get information. 

Not sure what the best way is to prevent any potential security issue there. The content of that file would be printed to anyone so it could have quite some impact.

Reply:
If `allow_url_fopen` then it is possible even that the file points to internal intranet services etc which could result in Server-Side Request Forgery .

Also I believe `fopen` might be disabled on some installations and would maybe need to fallback to something else. So far we're actually always using `file_get_contents` (which would also fetch data from URLs)

Reply:
I'll switch to using file_get_contents and add validation to make sure the path only starts with a slash or a dot and not a protocol.

If we're working off a limited supported set of mime type then recreating the image with imagecreatefromgif/jpeg/png should be fairly straightforward. I can do some performance tests to see how much it slows things down.

Reply:
I did a test using imagecreatefronstring() and imagepng() to recreate the image. Functionally it works fine, but performance is poor:

```
10,000x
image file 11kib PNG, regenerate   : 73.20s
 ... imagecreatefromstring()          8.09s
 ... imagepng()                      65.09s

image file 11kib PNG, pass through :  0.04s
image string 200b PNG, regenerate  :  0.17s
image string 200b PNG, pass through:  0.02s
```
All file loading is done outside the loop and there is no output being sent, so the time is all taken by the two GD functions.

11kib isn't an especially big image, so it seems that recreating the image on every response would create an unacceptable performance hit.

Reply:
I've reverted image recreation as I think it will be too slow.